rewrite-java/src/main/java/org/openrewrite/java/UsePrecompiledRegExpForReplaceAll.java

```diff
@@ -0,0 +1,117 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java;
+
+import org.openrewrite.Cursor;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.Space;
+import org.openrewrite.marker.Markers;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Pattern;
+
+import static org.openrewrite.Tree.randomId;
+
+public class UsePrecompiledRegExpForReplaceAll extends Recipe {
+
+    public static final String REG_EXP_KEY = "regExp";
+
+    @Override
+    public String getDisplayName() {
+        return "Replace `String.replaceAll` with a compiled regular expression";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Replace `String.replaceAll` with a compiled regular expression.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new ReplaceVisitor();
+    }
+
+    public static class ReplaceVisitor extends JavaIsoVisitor<ExecutionContext> {
+        // use boolean flag to only process one pattern at a time
+        private final AtomicBoolean patternProcessingInProgress = new AtomicBoolean(false);
+        private static final String PATTERN_VAR = "openRewriteReplaceAllPatternVar";
+        private static final JavaTemplate MATCHER_TEMPLATE = JavaTemplate
+                .builder("#{any()}.matcher( #{any(java.lang.CharSequence)}).replaceAll( #{any(java.lang.String)})")
+                .contextSensitive()
+                .imports("java.util.regex.Pattern")
+                .build();
+        private static final MethodMatcher REPLACE_ALL_METHOD_MATCHER = new MethodMatcher("java.lang.String replaceAll(String, String)");
+
+        @Override
+        public J.Block visitBlock(final J.Block block, final ExecutionContext executionContext) {
+            final J.Block processedBlock = super.visitBlock(block, executionContext);
+            final Cursor parentOrThrow = getCursor().getParentOrThrow();
+            if (parentOrThrow.getValue() instanceof J.ClassDeclaration) {
+                Object regExp = parentOrThrow.pollNearestMessage(REG_EXP_KEY);
+                if (regExp == null) {
+                    return processedBlock;
+                }
+                patternProcessingInProgress.set(false);
+
+                final String regExpName = parentOrThrow.pollNearestMessage(REG_EXP_KEY + "_NAME");
+                return super.visitBlock(
+                        JavaTemplate
+                                .builder("private static final java.util.regex.Pattern " + regExpName + " = Pattern.compile(#{});")
```

Comment:
Can you give me a hint on how to use substitution here for the regExpName? I was not able to get it working. It ended up with a weird content containing "private static final java.util.regex.Pattern __P__;/*__TEMPLATE_STOP__*/" and I couldn't get my head around it.